File: docs/algorithm/code/7. 整数反转/Code.dart
import 'dart:math';

/// 7. 整数反转
/// 给出一个 32 位的有符号整数，你需要将这个整数中每位上的数字进行反转。
/// 假设我们的环境只能存储得下 32 位的有符号整数，则其数值范围为 [−2^31,  2^31 − 1]。请根据这个假设，如果反转后整数溢出那么就返回 0。

final num MAX_VALUE = pow(2, 31) - 1;
final num MIN_VALUE = -pow(2, 31);
main() {
  print(MAX_VALUE);
  print(MIN_VALUE);
  // print(123);
  // print(reverse2(123));
  // print(-123);
  // print(reverse2(-123));
  // print(120);
  // print(reverse2(-120));
  // print(2147483647);
  // print(reverse2(7463847412));
}

int reverse2(int x) {
  int rev = 0;
  while (x != 0) {
    int pop = x % 10;
    x = (x ~/ 10).toInt();
    if (rev > MAX_VALUE / 10 || (rev == MAX_VALUE / 10 && pop > 7)) return 0;
    if (rev < MIN_VALUE / 10 || (rev == MIN_VALUE / 10 && pop < -8)) return 0;
    rev = rev * 10 + pop;
  }
  return rev;
}

int reverse(int x) {
  if (x > -10 && x < 10) {
    return x;
  }
  num subFlag = x < 0 ? -1 : 1;
  String str = (x * subFlag).toString();
  int length = str.length;
  StringBuffer buffer = new StringBuffer();
  while (length > 0) {
    buffer.write(str.substring(length - 1, length));
    length--;
  }

  num y = num.parse(buffer.toString()) * subFlag;
  if (y > MAX_VALUE || y < MIN_VALUE) return 0;
  return y.toInt();
}

File: docs/algorithm/simple/code/7. 整数反转/Code.java

/**
 * 7. 整数反转
 */
public class Code {
    public static void main(String[] args) {
        System.out.println(123);
        System.out.println(new Solution().reverse(123));
        System.out.println(-123);
        System.out.println(new Solution().reverse(-123));
        System.out.println(120);
        System.out.println(new Solution().reverse(120));
        System.out.println(1534236469);
        System.out.println(new Solution().reverse(1534236469));
        System.out.println(2147447412);
        System.out.println(new Solution().reverse(2147447412));
        System.out.println(new Solution().reverse2(2147447412));

    }
}

class Solution {
    public int reverse2(int x) {
        if (x > -10 && x < 10) {
            return x;
        }
        int rev = 0;
        while (x != 0) {
          int pop = x % 10;
          x = x / 10;
          if (rev > Integer.MAX_VALUE / 10 || (rev == Integer.MAX_VALUE / 10 && pop > 7)) return 0;
          if (rev < Integer.MIN_VALUE / 10 || (rev == Integer.MIN_VALUE / 10 && pop < -8)) return 0;
          rev = rev * 10 + pop;
        }
        return rev;
    }
    public int reverse(int x) {
        if (x > -10 && x < 10) {
            return x;
        }
        long subFlag = x < 0 ? -1l : 1l;
        StringBuffer buffer = new StringBuffer().append(x * subFlag);
        long y = Long.valueOf(buffer.reverse().toString()) * subFlag;
        if (y > Integer.MAX_VALUE || y < Integer.MIN_VALUE)
            return 0;
        return (int) (y);
    }
}
